unit_tests/log/test_main_chkpt_info.cpp

```diff
@@ -0,0 +1,960 @@
+  env.get_after ()->unpack (unpac);
+
+  env.require_equal (*env.get_before (), *env.get_after ());
+}
+
+TEST_CASE ("Test pack/unpack checkpoint_info class 2", "")
+{
+  test_env_chkpt env {0, 100};
+
+  cubpacking::packer pac;
+  size_t size = env.get_before ()->get_packed_size (pac, 0);
+  char buffer[size];
+  pac.set_buffer (buffer, size);
+  env.get_before ()->pack (pac);
+
+  cubpacking::unpacker unpac;
+  unpac.set_buffer (buffer, size);
+  env.get_after ()->unpack (unpac);
+
+  env.require_equal (*env.get_before (), *env.get_after ());
+}
+
+TEST_CASE ("Test pack/unpack checkpoint_info class 3", "")
+{
+  test_env_chkpt env {220, 80};
+
+  cubpacking::packer pac;
+  size_t size = env.get_before ()->get_packed_size (pac, 0);
+  size += env.get_before ()->get_packed_size (pac, size);
+  char buffer[size];
+  pac.set_buffer (buffer, size);
+  env.get_before ()->pack (pac);
+  env.get_before ()->pack (pac);
+
+  cubpacking::unpacker unpac;
+  unpac.set_buffer (buffer, size);
+  env.get_after ()->unpack (unpac);
+  env.require_equal (*env.get_before (), *env.get_after ());
+
+  checkpoint_info after_2;
+
+  after_2.unpack (unpac);
+
+  env.require_equal (*env.get_after (), after_2);
+}
+
+std::map<TRANID, log_tdes *> tran_map;
+std::map<TRANID, log_tdes *> systb_System_tdes;
+int count_2pc_after_recovery;
+
+TEST_CASE ("Test load and recovery on empty tran table", "")
+{
+  test_env_chkpt env;
+  LOG_LSA smallest_lsa;
+  LOG_LSA star_lsa;
+  THREAD_ENTRY thd;
+
+  env.get_after ()->load_trantable_snapshot (&thd, smallest_lsa);
+  env.get_after ()->recovery_analysis (&thd, star_lsa);
+  env.get_after ()->recovery_2pc_analysis (&thd);
+  REQUIRE (tran_map.size() == 0);
+}
+
+int
+search_for_id (TRANID id)
+{
+  for (int i = 0; i < log_Gl.trantable.num_total_indices; i++)
+    {
+      if (log_Gl.trantable.all_tdes[i]->trid == id)
+	{
+	  return 1;
+	}
+    }
+  return 0;
+}
+
+void
+full_compare_tdes (log_tdes *td1, const log_tdes *td2)
+{
+  REQUIRE (td1->head_lsa == td2->head_lsa);
+  REQUIRE (td1->tail_lsa == td2->tail_lsa);
+  REQUIRE (td1->undo_nxlsa == td2->undo_nxlsa);
+  REQUIRE (td1->posp_nxlsa == td2->posp_nxlsa);
+  REQUIRE (td1->savept_lsa == td2->savept_lsa);
+  REQUIRE (td1->tail_topresult_lsa == td2->tail_topresult_lsa);
+  REQUIRE (td1->rcv.sysop_start_postpone_lsa == td2->rcv.sysop_start_postpone_lsa);
+  REQUIRE (td1->rcv.atomic_sysop_start_lsa == td2->rcv.atomic_sysop_start_lsa);
+}
+
+void
+full_compare_tdes_active (log_tdes *td1, const log_tdes *td2)
+{
+  REQUIRE (td1->head_lsa == td2->head_lsa);
+  REQUIRE (td1->tail_lsa == td2->undo_nxlsa);
+  REQUIRE (td1->tail_lsa == td2->tail_lsa);
+  REQUIRE (td1->posp_nxlsa == td2->posp_nxlsa);
+  REQUIRE (td1->savept_lsa == td2->savept_lsa);
+  REQUIRE (td1->tail_topresult_lsa == td2->tail_topresult_lsa);
+  REQUIRE (td1->rcv.sysop_start_postpone_lsa == td2->rcv.sysop_start_postpone_lsa);
+  REQUIRE (td1->rcv.atomic_sysop_start_lsa == td2->rcv.atomic_sysop_start_lsa);
+}
```

Comment:
undo_nxlsa after analysis is equal to either tdes_before.undo_nxlsa or to tdes_before.tail_lsa (see checkpoint_info::load_checkpoint_trans). Is this modification enough to remove full_compare_tdes_active?

```suggestion
void
full_compare_tdes (log_tdes *td_before, const log_tdes *td_after)
{
  REQUIRE (td_before->head_lsa == td_after->head_lsa);
  REQUIRE (td_before->tail_lsa == td_after->tail_lsa);
  REQUIRE (td_before->undo_nxlsa == td_after->undo_nxlsa || td_before->tail_lsa == td_after->undo_nxlsa);
  REQUIRE (td_before->posp_nxlsa == td_after->posp_nxlsa);
  REQUIRE (td_before->savept_lsa == td_after->savept_lsa);
  REQUIRE (td_before->tail_topresult_lsa == td_after->tail_topresult_lsa);
  REQUIRE (td_before->rcv.sysop_start_postpone_lsa == td_after->rcv.sysop_start_postpone_lsa);
  REQUIRE (td_before->rcv.atomic_sysop_start_lsa == td_after->rcv.atomic_sysop_start_lsa);
}
```

Reply:
chained comparisons are not permited, fixed by adding a new param to the full_compare_tdes function